File: Reaktoro/Singletons/Elements.py
# Reaktoro is a unified framework for modeling chemically reactive systems.
#
# Copyright © 2014-2022 Allan Leal
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library. If not, see <http://www.gnu.org/licenses/>.


from reaktoro import *


def testElements():

    assert Elements.size() == len(Elements.data())

    assert Elements.withSymbol("H"  ).name() == "Hydrogen"
    assert Elements.withSymbol("He" ).name() == "Helium"
    assert Elements.withSymbol("Li" ).name() == "Lithium"
    assert Elements.withSymbol("Be" ).name() == "Beryllium"
    assert Elements.withSymbol("B"  ).name() == "Boron"
    assert Elements.withSymbol("C"  ).name() == "Carbon"
    assert Elements.withSymbol("N"  ).name() == "Nitrogen"
    assert Elements.withSymbol("O"  ).name() == "Oxygen"
    assert Elements.withSymbol("F"  ).name() == "Fluorine"
    assert Elements.withSymbol("Ne" ).name() == "Neon"
    assert Elements.withSymbol("Na" ).name() == "Sodium"
    assert Elements.withSymbol("Mg" ).name() == "Magnesium"
    assert Elements.withSymbol("Al" ).name() == "Aluminum"
    assert Elements.withSymbol("Si" ).name() == "Silicon"
    assert Elements.withSymbol("P"  ).name() == "Phosphorus"
    assert Elements.withSymbol("S"  ).name() == "Sulfur"
    assert Elements.withSymbol("Cl" ).name() == "Chlorine"
    assert Elements.withSymbol("Ar" ).name() == "Argon"
    assert Elements.withSymbol("K"  ).name() == "Potassium"
    assert Elements.withSymbol("Ca" ).name() == "Calcium"
    assert Elements.withSymbol("Sc" ).name() == "Scandium"
    assert Elements.withSymbol("Ti" ).name() == "Titanium"
    assert Elements.withSymbol("V"  ).name() == "Vanadium"
    assert Elements.withSymbol("Cr" ).name() == "Chromium"
    assert Elements.withSymbol("Mn" ).name() == "Manganese"
    assert Elements.withSymbol("Fe" ).name() == "Iron"
    assert Elements.withSymbol("Co" ).name() == "Cobalt"
    assert Elements.withSymbol("Ni" ).name() == "Nickel"
    assert Elements.withSymbol("Cu" ).name() == "Copper"
    assert Elements.withSymbol("Zn" ).name() == "Zinc"
    assert Elements.withSymbol("Ga" ).name() == "Gallium"
    assert Elements.withSymbol("Ge" ).name() == "Germanium"
    assert Elements.withSymbol("As" ).name() == "Arsenic"
    assert Elements.withSymbol("Se" ).name() == "Selenium"
    assert Elements.withSymbol("Br" ).name() == "Bromine"
    assert Elements.withSymbol("Kr" ).name() == "Krypton"
    assert Elements.withSymbol("Rb" ).name() == "Rubidium"
    assert Elements.withSymbol("Sr" ).name() == "Strontium"
    assert Elements.withSymbol("Y"  ).name() == "Yttrium"
    assert Elements.withSymbol("Zr" ).name() == "Zirconium"
    assert Elements.withSymbol("Nb" ).name() == "Niobium"
    assert Elements.withSymbol("Mo" ).name() == "Molybdenum"
    assert Elements.withSymbol("Tc" ).name() == "Technetium"
    assert Elements.withSymbol("Ru" ).name() == "Ruthenium"
    assert Elements.withSymbol("Rh" ).name() == "Rhodium"
    assert Elements.withSymbol("Pd" ).name() == "Palladium"
    assert Elements.withSymbol("Ag" ).name() == "Silver"
    assert Elements.withSymbol("Cd" ).name() == "Cadmium"
    assert Elements.withSymbol("In" ).name() == "Indium"
    assert Elements.withSymbol("Sn" ).name() == "Tin"
    assert Elements.withSymbol("Sb" ).name() == "Antimony"
    assert Elements.withSymbol("Te" ).name() == "Tellurium"
    assert Elements.withSymbol("I"  ).name() == "Iodine"
    assert Elements.withSymbol("Xe" ).name() == "Xenon"
    assert Elements.withSymbol("Cs" ).name() == "Cesium"
    assert Elements.withSymbol("Ba" ).name() == "Barium"
    assert Elements.withSymbol("La" ).name() == "Lanthanum"
    assert Elements.withSymbol("Ce" ).name() == "Cerium"
    assert Elements.withSymbol("Pr" ).name() == "Praseodymium"
    assert Elements.withSymbol("Nd" ).name() == "Neodymium"
    assert Elements.withSymbol("Pm" ).name() == "Promethium"
    assert Elements.withSymbol("Sm" ).name() == "Samarium"
    assert Elements.withSymbol("Eu" ).name() == "Europium"
    assert Elements.withSymbol("Gd" ).name() == "Gadolinium"
    assert Elements.withSymbol("Tb" ).name() == "Terbium"
    assert Elements.withSymbol("Dy" ).name() == "Dysprosium"
    assert Elements.withSymbol("Ho" ).name() == "Holmium"
    assert Elements.withSymbol("Er" ).name() == "Erbium"
    assert Elements.withSymbol("Tm" ).name() == "Thulium"
    assert Elements.withSymbol("Yb" ).name() == "Ytterbium"
    assert Elements.withSymbol("Lu" ).name() == "Lutetium"
    assert Elements.withSymbol("Hf" ).name() == "Hafnium"
    assert Elements.withSymbol("Ta" ).name() == "Tantalum"
    assert Elements.withSymbol("W"  ).name() == "Tungsten"
    assert Elements.withSymbol("Re" ).name() == "Rhenium"
    assert Elements.withSymbol("Os" ).name() == "Osmium"
    assert Elements.withSymbol("Ir" ).name() == "Iridium"
    assert Elements.withSymbol("Pt" ).name() == "Platinum"
    assert Elements.withSymbol("Au" ).name() == "Gold"
    assert Elements.withSymbol("Hg" ).name() == "Mercury"
    assert Elements.withSymbol("Tl" ).name() == "Thallium"
    assert Elements.withSymbol("Pb" ).name() == "Lead"
    assert Elements.withSymbol("Bi" ).name() == "Bismuth"
    assert Elements.withSymbol("Po" ).name() == "Polonium"
    assert Elements.withSymbol("At" ).name() == "Astatine"
    assert Elements.withSymbol("Rn" ).name() == "Radon"
    assert Elements.withSymbol("Fr" ).name() == "Francium"
    assert Elements.withSymbol("Ra" ).name() == "Radium"
    assert Elements.withSymbol("Ac" ).name() == "Actinium"
    assert Elements.withSymbol("Th" ).name() == "Thorium"
    assert Elements.withSymbol("Pa" ).name() == "Protactinium"
    assert Elements.withSymbol("U"  ).name() == "Uranium"
    assert Elements.withSymbol("Np" ).name() == "Neptunium"
    assert Elements.withSymbol("Pu" ).name() == "Plutonium"
    assert Elements.withSymbol("Am" ).name() == "Americium"
    assert Elements.withSymbol("Cm" ).name() == "Curium"
    assert Elements.withSymbol("Bk" ).name() == "Berkelium"
    assert Elements.withSymbol("Cf" ).name() == "Californium"
    assert Elements.withSymbol("Es" ).name() == "Einsteinium"
    assert Elements.withSymbol("Fm" ).name() == "Fermium"
    assert Elements.withSymbol("Md" ).name() == "Mendelevium"
    assert Elements.withSymbol("No" ).name() == "Nobelium"
    assert Elements.withSymbol("Lr" ).name() == "Lawrencium"
    assert Elements.withSymbol("Rf" ).name() == "Rutherfordium"
    assert Elements.withSymbol("Db" ).name() == "Dubnium"
    assert Elements.withSymbol("Sg" ).name() == "Seaborgium"
    assert Elements.withSymbol("Bh" ).name() == "Bohrium"
    assert Elements.withSymbol("Hs" ).name() == "Hassium"
    assert Elements.withSymbol("Mt" ).name() == "Meitnerium"
    assert Elements.withSymbol("Ds" ).name() == "Darmstadtium"
    assert Elements.withSymbol("Rg" ).name() == "Roentgenium"
    assert Elements.withSymbol("Cn" ).name() == "Copernicium"
    assert Elements.withSymbol("Nh" ).name() == "Nihonium"
    assert Elements.withSymbol("Fl" ).name() == "Flerovium"
    assert Elements.withSymbol("Mc" ).name() == "Moscovium"
    assert Elements.withSymbol("Lv" ).name() == "Livermorium"
    assert Elements.withSymbol("Ts" ).name() == "Tennessine"
    assert Elements.withSymbol("Og" ).name() == "Oganesson"

    assert Elements.withName("Hydrogen").symbol() == "H"
    assert Elements.withName("Helium"  ).symbol() == "He"
    assert Elements.withName("Lithium" ).symbol() == "Li"

    assert Elements.withSymbol("Ab") == None
    assert Elements.withSymbol("Xy") == None

    Elements.append(Element().withSymbol("Aa").withTags(["tag1", "tag2", "tag3"]))
    Elements.append(Element().withSymbol("Bb").withTags(["tag1", "tag3"]))
    Elements.append(Element().withSymbol("Cc").withTags(["tag1"]))

    elements_with_tag = Elements.withTag("tag1")

    assert len(elements_with_tag) == 3
    assert elements_with_tag[0].symbol() == "Aa"
    assert elements_with_tag[1].symbol() == "Bb"
    assert elements_with_tag[2].symbol() == "Cc"

    elements_with_tags = Elements.withTags(["tag3", "tag1"])

    assert len(elements_with_tags) == 2
    assert elements_with_tags[0].symbol() == "Aa"
    assert elements_with_tags[1].symbol() == "Bb"

File: Reaktoro/Singletons/Elements.cpp
// Reaktoro is a unified framework for modeling chemically reactive systems.
//
// Copyright © 2014-2024 Allan Leal
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this library. If not, see <http://www.gnu.org/licenses/>.

#include "Elements.hpp"

// Reaktoro includes
#include <Reaktoro/Common/Algorithms.hpp>
#include <Reaktoro/Common/Exception.hpp>

namespace Reaktoro {
namespace detail {

/// The default elements for the Elements object.
const Vec<Element> default_elements =
{
    Element({ "H"  , 0.001007940 , "Hydrogen"      }),
    Element({ "He" , 0.004002602 , "Helium"        }),
    Element({ "Li" , 0.006941000 , "Lithium"       }),
    Element({ "Be" , 0.009012180 , "Beryllium"     }),
    Element({ "B"  , 0.010811000 , "Boron"         }),
    Element({ "C"  , 0.012011000 , "Carbon"        }),
    Element({ "N"  , 0.014006740 , "Nitrogen"      }),
    Element({ "O"  , 0.015999400 , "Oxygen"        }),
    Element({ "F"  , 0.018998403 , "Fluorine"      }),
    Element({ "Ne" , 0.020179700 , "Neon"          }),
    Element({ "Na" , 0.022989768 , "Sodium"        }),
    Element({ "Mg" , 0.024305000 , "Magnesium"     }),
    Element({ "Al" , 0.026981539 , "Aluminum"      }),
    Element({ "Si" , 0.028085500 , "Silicon"       }),
    Element({ "P"  , 0.030973762 , "Phosphorus"    }),
    Element({ "S"  , 0.032066000 , "Sulfur"        }),
    Element({ "Cl" , 0.035452700 , "Chlorine"      }),
    Element({ "Ar" , 0.039948000 , "Argon"         }),
    Element({ "K"  , 0.039098300 , "Potassium"     }),
    Element({ "Ca" , 0.040078000 , "Calcium"       }),
    Element({ "Sc" , 0.044955910 , "Scandium"      }),
    Element({ "Ti" , 0.047880000 , "Titanium"      }),
    Element({ "V"  , 0.050941500 , "Vanadium"      }),
    Element({ "Cr" , 0.051996100 , "Chromium"      }),
    Element({ "Mn" , 0.054938050 , "Manganese"     }),
    Element({ "Fe" , 0.055847000 , "Iron"          }),
    Element({ "Co" , 0.058933200 , "Cobalt"        }),
    Element({ "Ni" , 0.058693400 , "Nickel"        }),
    Element({ "Cu" , 0.063546000 , "Copper"        }),
    Element({ "Zn" , 0.065390000 , "Zinc"          }),
    Element({ "Ga" , 0.069723000 , "Gallium"       }),
    Element({ "Ge" , 0.072610000 , "Germanium"     }),
    Element({ "As" , 0.074921590 , "Arsenic"       }),
    Element({ "Se" , 0.078960000 , "Selenium"      }),
    Element({ "Br" , 0.079904000 , "Bromine"       }),
    Element({ "Kr" , 0.083800000 , "Krypton"       }),
    Element({ "Rb" , 0.085467800 , "Rubidium"      }),
    Element({ "Sr" , 0.087620000 , "Strontium"     }),
    Element({ "Y"  , 0.088905850 , "Yttrium"       }),
    Element({ "Zr" , 0.091224000 , "Zirconium"     }),
    Element({ "Nb" , 0.092906380 , "Niobium"       }),
    Element({ "Mo" , 0.095940000 , "Molybdenum"    }),
    Element({ "Tc" , 0.097907200 , "Technetium"    }),
    Element({ "Ru" , 0.101070000 , "Ruthenium"     }),
    Element({ "Rh" , 0.102905500 , "Rhodium"       }),
    Element({ "Pd" , 0.106420000 , "Palladium"     }),
    Element({ "Ag" , 0.107868200 , "Silver"        }),
    Element({ "Cd" , 0.112411000 , "Cadmium"       }),
    Element({ "In" , 0.114818000 , "Indium"        }),
    Element({ "Sn" , 0.118710000 , "Tin"           }),
    Element({ "Sb" , 0.121760000 , "Antimony"      }),
    Element({ "Te" , 0.127600000 , "Tellurium"     }),
    Element({ "I"  , 0.126904470 , "Iodine"        }),
    Element({ "Xe" , 0.131290000 , "Xenon"         }),
    Element({ "Cs" , 0.132905430 , "Cesium"        }),
    Element({ "Ba" , 0.137327000 , "Barium"        }),
    Element({ "La" , 0.138905500 , "Lanthanum"     }),
    Element({ "Ce" , 0.140115000 , "Cerium"        }),
    Element({ "Pr" , 0.140907650 , "Praseodymium"  }),
    Element({ "Nd" , 0.144240000 , "Neodymium"     }),
    Element({ "Pm" , 0.144912700 , "Promethium"    }),
    Element({ "Sm" , 0.150360000 , "Samarium"      }),
    Element({ "Eu" , 0.151965000 , "Europium"      }),
    Element({ "Gd" , 0.157250000 , "Gadolinium"    }),
    Element({ "Tb" , 0.158925340 , "Terbium"       }),
    Element({ "Dy" , 0.162500000 , "Dysprosium"    }),
    Element({ "Ho" , 0.164930320 , "Holmium"       }),
    Element({ "Er" , 0.167260000 , "Erbium"        }),
    Element({ "Tm" , 0.168934210 , "Thulium"       }),
    Element({ "Yb" , 0.173040000 , "Ytterbium"     }),
    Element({ "Lu" , 0.174967000 , "Lutetium"      }),
    Element({ "Hf" , 0.178490000 , "Hafnium"       }),
    Element({ "Ta" , 0.180947900 , "Tantalum"      }),
    Element({ "W"  , 0.183840000 , "Tungsten"      }),
    Element({ "Re" , 0.186207000 , "Rhenium"       }),
    Element({ "Os" , 0.190230000 , "Osmium"        }),
    Element({ "Ir" , 0.192220000 , "Iridium"       }),
    Element({ "Pt" , 0.195080000 , "Platinum"      }),
    Element({ "Au" , 0.196966540 , "Gold"          }),
    Element({ "Hg" , 0.200590000 , "Mercury"       }),
    Element({ "Tl" , 0.204383300 , "Thallium"      }),
    Element({ "Pb" , 0.207200000 , "Lead"          }),
    Element({ "Bi" , 0.208980370 , "Bismuth"       }),
    Element({ "Po" , 0.208982400 , "Polonium"      }),
    Element({ "At" , 0.209987100 , "Astatine"      }),
    Element({ "Rn" , 0.222017600 , "Radon"         }),
    Element({ "Fr" , 0.223019700 , "Francium"      }),
    Element({ "Ra" , 0.226025400 , "Radium"        }),
    Element({ "Ac" , 0.227027800 , "Actinium"      }),
    Element({ "Th" , 0.232038100 , "Thorium"       }),
    Element({ "Pa" , 0.231035880 , "Protactinium"  }),
    Element({ "U"  , 0.238028900 , "Uranium"       }),
    Element({ "Np" , 0.237048000 , "Neptunium"     }),
    Element({ "Pu" , 0.244064200 , "Plutonium"     }),
    Element({ "Am" , 0.243061400 , "Americium"     }),
    Element({ "Cm" , 0.247070300 , "Curium"        }),
    Element({ "Bk" , 0.247070300 , "Berkelium"     }),
    Element({ "Cf" , 0.251079600 , "Californium"   }),
    Element({ "Es" , 0.252083000 , "Einsteinium"   }),
    Element({ "Fm" , 0.257095100 , "Fermium"       }),
    Element({ "Md" , 0.258100000 , "Mendelevium"   }),
    Element({ "No" , 0.259100900 , "Nobelium"      }),
    Element({ "Lr" , 0.262110000 , "Lawrencium"    }),
    Element({ "Rf" , 0.261000000 , "Rutherfordium" }),
    Element({ "Db" , 0.262000000 , "Dubnium"       }),
    Element({ "Sg" , 0.266000000 , "Seaborgium"    }),
    Element({ "Bh" , 0.264000000 , "Bohrium"       }),
    Element({ "Hs" , 0.269000000 , "Hassium"       }),
    Element({ "Mt" , 0.268000000 , "Meitnerium"    }),
    Element({ "Ds" , 0.269000000 , "Darmstadtium"  }),
    Element({ "Rg" , 0.272000000 , "Roentgenium"   }),
    Element({ "Cn" , 0.277000000 , "Copernicium"   }),
    Element({ "Nh" , 0.000000000 , "Nihonium"      }),
    Element({ "Fl" , 0.289000000 , "Flerovium"     }),
    Element({ "Mc" , 0.000000000 , "Moscovium"     }),
    Element({ "Lv" , 0.000000000 , "Livermorium"   }),
    Element({ "Ts" , 0.000000000 , "Tennessine"    }),
    Element({ "Og" , 0.000000000 , "Oganesson"     }),
    Element({ "D"  , 0.002014102 , "Deuterium"     }),
    Element({ "T"  , 0.003016049 , "Tritium"       }),
};

} // namespace detail

Elements::Elements()
: m_elements(detail::default_elements)
{}

Elements::~Elements()
{}

auto Elements::instance() -> Elements&
{
    static Elements obj;
    return obj;
}

auto Elements::data() -> Vec<Element> const&
{
    return instance().m_elements;
}

auto Elements::elements() -> Vec<Element> const&
{
    return instance().m_elements;
}

auto Elements::append(Element element) -> void
{
    auto& elements = instance().m_elements;
    elements.emplace_back(std::move(element));
}

auto Elements::size() -> std::size_t
{
    return data().size();
}

auto Elements::withSymbol(String symbol) -> Optional<Element>
{
    const auto idx = indexfn(data(), [&](auto&& e) { return e.symbol() == symbol; });
    if(idx < size()) return data()[idx];
    return {};
}

auto Elements::withName(String name) -> Optional<Element>
{
    const auto idx = indexfn(data(), [&](auto&& e) { return e.name() == name; });
    if(idx < size()) return data()[idx];
    return {};
}

auto Elements::withTag(String tag) -> Vec<Element>
{
    return filter(data(), [&](auto&& e) { return contains(e.tags(), tag); });
}

auto Elements::withTags(StringList const& tags) -> Vec<Element>
{
    return filter(data(), [&](auto&& e) { return contained(tags, e.tags()); });
}

} // namespace Reaktoro
